lib/TPP/GPU/LinalgToXeGPU.cpp

```diff
@@ -0,0 +1,1394 @@
+  auto vecLoadType = cast<VectorType>(loadVecTiles[0].getType());
+  assert(llvm::all_of(loadVecTiles,
+                      [&](Value tile) {
+                        return cast<VectorType>(tile.getType()) == vecLoadType;
+                      }) &&
+         "All loaded vectors must have the same type.");
+  assert(vecLoadType.getShape().size() == 2 ||
+         vnniConf && "Requires VNNI config for non 2D loaded tiles");
+
+  // Accumulate all dimensions as the vector might have extra VNNI
+  // dimensions.
+  int loadVecSize = std::accumulate(vecLoadType.getShape().begin(),
+                                    vecLoadType.getShape().end(), 1,
+                                    std::multiplies<int64_t>());
+  auto loadVecFlat = VectorType::get(loadVecSize, vecLoadType.getElementType());
+
+  VectorType vecSubTileType =
+      VectorType::get(subTileShape, vecLoadType.getElementType());
+  if (vnniConf) {
+    vecSubTileType =
+        getVnniVector(subTileShape, vecLoadType.getElementType(), *vnniConf);
+  }
+
+  const int totalTileRows = totalTileShape[0] / loadTileShape[0];
+  const int totalTileCols = totalTileShape[1] / loadTileShape[1];
+
+  const int subTilesPerLoadRow = loadTileShape[0] / subTileShape[0];
+  const int subTilePerLoadCol = loadTileShape[1] / subTileShape[1];
+
+  const int subTileRows = totalTileShape[0] / subTileShape[0];
+  const int subTileCols = totalTileShape[1] / subTileShape[1];
+  TilesArray subTiles(subTileRows, subTileCols);
+
+  // Iterate over the total tile.
+  for (int m = 0; m < totalTileRows; m++) {
+    for (int k = 0; k < totalTileCols; k++) {
+      // Load tiles are ordered in row-major fashion.
+      int loadIdx = m * totalTileCols + k;
+      auto totalTileShape = loadVecTiles[loadIdx];
+      auto castFlat = rewriter.create<vector::ShapeCastOp>(loc, loadVecFlat,
+                                                           totalTileShape);
+
+      // Iterate over load tiles.
+      // Each load tile contains one or more sub-tiles.
+      for (int i = 0; i < subTilesPerLoadRow; i++) {
+        for (int j = 0; j < subTilePerLoadCol; j++) {
+          const int subTileShapeSize = subTileShape[0] * subTileShape[1];
+          int dpasIdx = i * subTilePerLoadCol + j;
+          int offset = dpasIdx * subTileShapeSize;
+
+          auto slice = rewriter.create<vector::ExtractStridedSliceOp>(
+              loc, castFlat, /*offsets=*/ArrayRef<int64_t>{offset},
+              /*sizes=*/ArrayRef<int64_t>{subTileShapeSize},
+              /*strides=*/ArrayRef<int64_t>{1});
+          auto castTile =
+              rewriter.create<vector::ShapeCastOp>(loc, vecSubTileType, slice);
+
+          // Insert the sub-tiles in their position relative to the whole
+          // thread/workitem tile.
+          int rowIdx = m * subTilesPerLoadRow + i;
+          int colIdx = k * subTilePerLoadCol + j;
+          subTiles.setTile(rowIdx, colIdx, castTile);
```

Comment:
what happens if rowIdx and colIdx are larger than 8 here? The TilesArray will automatically expand?

Reply:
At this point `subTiles` is just a 2D array, its constructor ensures the storage is as large as requested which is known as all relevant sizes are statically known. But it won't resize any further.